pkg/kubelet/eviction/eviction_manager_test.go

```diff
@@ -3108,3 +2976,135 @@ func TestManagerWithLocalStorageCapacityIsolationOpen(t *testing.T) {
 		t.Fatalf("Unexpected evicted pod (-want,+got):\n%s", diff)
 	}
 }
+
+// TestHardEvictPodThatHasBeenSoftEvictedOnceHardThresholdReached verifies that if a pod is taking too long to evict,
+// and we have reached the hard eviction threshold, we proceed to hard evict it
+func TestHardEvictPodThatHasBeenSoftEvictedOnceHardThresholdReached(t *testing.T) {
+	podMaker := makePodWithMemoryStats
+	summaryStatsMaker := makeMemoryStats
+	podsToMake := []podToMake{
+		{name: "this-one-goes", priority: defaultPriority, requests: newResourceList("100m", "1Gi", ""), limits: newResourceList("100m", "1Gi", ""), memoryWorkingSet: "1Gi"},
+		{name: "this-one-stays", priority: defaultPriority, requests: newResourceList("100m", "1Gi", ""), limits: newResourceList("100m", "1Gi", ""), memoryWorkingSet: "900Mi"},
+		{name: "this-one-stays-too", priority: defaultPriority, requests: newResourceList("100m", "1Gi", ""), limits: newResourceList("100m", "1Gi", ""), memoryWorkingSet: "100Mi"},
+	}
+	pods := []*v1.Pod{}
+	podStats := map[*v1.Pod]statsapi.PodStats{}
+	for _, podToMake := range podsToMake {
+		pod, podStat := podMaker(podToMake.name, podToMake.priority, podToMake.requests, podToMake.limits, podToMake.memoryWorkingSet)
+		pods = append(pods, pod)
+		podStats[pod] = podStat
+	}
+	podToEvict := pods[0]
+	activePodsFunc := func() []*v1.Pod {
+		return pods
+	}
+
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	podKiller := &mockPodKiller{}
+	diskInfoProvider := &mockDiskInfoProvider{dedicatedImageFs: ptr.To(false)}
+	nodeRef := &v1.ObjectReference{Kind: "Node", Name: "test", UID: types.UID("test"), Namespace: ""}
+
+	config := Config{
+		MaxPodGracePeriodSeconds: 5,
+		PressureTransitionPeriod: time.Minute * 5,
+		Thresholds: []evictionapi.Threshold{
+			{ // soft
+				Signal:   evictionapi.SignalMemoryAvailable,
+				Operator: evictionapi.OpLessThan,
+				Value: evictionapi.ThresholdValue{
+					Quantity: quantityMustParse("1Gi"),
+				},
+				GracePeriod: 10,
+			},
+			{ // hard
+				Signal:   evictionapi.SignalMemoryAvailable,
+				Operator: evictionapi.OpLessThan,
+				Value: evictionapi.ThresholdValue{
+					Quantity: quantityMustParse("500Mi"),
+				},
+				GracePeriod: 0,
+			},
+		},
+	}
+	summaryProvider := &fakeSummaryProvider{result: summaryStatsMaker("600Mi", podStats)}
+	manager := newManagerImpl(fakeClock, podKiller.killPodNow, config, summaryProvider, nodeRef)
+	manager.killPodFunc = podKiller.killPodNowLongShutdown
+	fakeClock.Step(1 * time.Minute)
+
+	// first run doesn't meet the grace period
+	_, err := manager.synchronize(diskInfoProvider, activePodsFunc)
+	if err != nil {
+		t.Fatalf("Manager should not report any errors")
+	}
+	if podKiller.pod != nil {
+		t.Fatalf("Manager should not have chosen to kill a pod, but it did: %v", podKiller.pod.Name)
+	}
+
+	// now we meet the grace period of soft eviction
+	fakeClock.Step(1 * time.Minute)
```

Comment:
I'm missing something here.

The grace period is defined as `10` (or 5 by MaxPodGracePeriodSeconds), so why stepping 1 minute shouldn't meet this period? Why stepping another minute is expected to meet it? What am I missing?

Reply:
Ah yeah, this is confusing. The 10 is actually 10ns, so we don't need the minute sleep, a lot less will do